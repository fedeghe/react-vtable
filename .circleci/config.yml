version: 2
jobs:
  test:
    docker:
      - image: circleci/node:12.16
    steps:
      - checkout
      - run: echo "Running tests"
      - run: yarn
      - run: yarn test
  build:
    docker:
      - image: circleci/node:12.16
    steps:
      - checkout
      - run: echo "build project"
      - run: yarn
      - run: yarn build

workflows:
  version: 2
    test_build:
      jobs:
        - test
        - build:
          requires:
            - test