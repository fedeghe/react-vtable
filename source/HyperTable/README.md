# caption

css
caption-side: top, bottom
text-align: left, center, right

attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes

# col

attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes
- span

# colgroup

attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes
- span

# table

# tbody

# td
attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes
- colspan
- headers
- rowspan

# tfoot

# th
attributes:
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes
- abbr
- colspan
- headers
- rowspan
- scope

# thead
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes

# tr
https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes





