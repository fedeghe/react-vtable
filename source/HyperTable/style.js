import { createUseStyles } from "react-jss";

export default createUseStyles(() => ({
    Wrapper : ({width, height}) => ({
        height,
        width,
        backgroundColor: 'white',
        overflow: 'hidden',
        position:'relative'
    }),
}));