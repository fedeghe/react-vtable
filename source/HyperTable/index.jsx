import React, { useReducer, useEffect, useRef} from 'react'

import reducerFactory from './reducer'
import { replaceall } from './utils'

import './style.less'

const HyperTable = ({
    columns,
    data,
    captionTop,
    captionBottom,
    height, width,
    rowVerticalAlign,
    crossHighlight,
    columnHighlight,
    rowHighlight,
    cellHightlight,
    noFilterData,
    columnClick = () => { },
    columnEnter = () => { },
    columnLeave = () => { },
    rowClick = () => { },
    rowEnter = () => { },
    rowLeave = () => { },
    cellClick = () => { },
    cellEnter = () => { },
    cellLeave = () => { },
}) => {
    const { reducer, init } = reducerFactory()
    const [state, dispatch] = useReducer(reducer, init(data))
    const {total, rows, activeRow, activeCol, filters, sorting : {
        field: sortingField,
        versus: sortingVersus
    } = {}} = state
    const table = useRef(null);

    

    useEffect(() => {
        table.current.style.display = 'table';
        return () => {
            table.current.style.display = 'none'
        }
    }, [])

    return <div className="TableWrapper">
        <table ref={table} style={{ width: width }} className="Table" border="0" cellSpacing="0" >
            {captionTop && (
                <caption
                    style={{ 'captionSide': 'top'}}
                    className={`TableCaption ${captionTop.className}`}
                >{
                    'component' in captionTop
                    ? captionTop.component({
                        count: rows.length,
                        total, filters,sorting
                    })
                    : replaceall(captionTop.text, {COUNT: rows.length, TOTAL: total})
                }</caption>
            )}
            {captionBottom && (
                <caption
                    style={{ 'captionSide': 'bottom' }}
                    className={`TableCaption ${captionBottom.className}`}
                >{
                    'component' in captionBottom
                    ? captionBottom.component({
                        count: rows.length,
                        total, filters, sorting
                    })
                    : replaceall(captionBottom.text, {COUNT: rows.length, TOTAL: total})
                }</caption>
            )}
            <thead className="TableHeader ">
                <tr>{
                    columns.map((col, k) => (
                        <th key={`h${k}`}
                            className={`TableHeaderCell tablecell`}
                            style={col.width && { width: col.width }}
                        >
                            <div className="tableheaderwrapper">
                                <div>{
                                    'headerComponent' in col
                                        ? col.headerComponent(col, 'key')
                                        : col.headerLabel || col.key
                                }</div>
                                <div className="tableheaderoptions">{
                                    col.filter && col.filter(value => dispatch({
                                        type: 'filter',
                                        payload: {
                                            field: col.key,
                                            value
                                        }
                                    }))
                                }
                                {
                                    col.sorting && (
                                        <div className="tableheadercellsort">
                                            <div className={(sortingField === col.key && sortingVersus === 1) ? 'sortActive' : 'sortInactive'} onClick={() =>
                                                dispatch({
                                                    type: 'sortBy',
                                                    payload: {
                                                        sort: col.sorting.sort(1),
                                                        field: col.key,
                                                        versus: 1
                                                    }
                                                })
                                            }>▲</div>
                                            <div className={(sortingField === null) ? 'sortActive' : 'sortInactive'} onClick={() =>
                                                dispatch({
                                                    type: 'unsortBy',
                                                })
                                            }>&bull;</div>
                                            <div className={(sortingField === col.key && sortingVersus === -1) ? 'sortActive' : 'sortInactive'} onClick={() =>
                                                dispatch({
                                                    type: 'sortBy',
                                                    payload: {
                                                        sort: col.sorting.sort(-1),
                                                        field: col.key,
                                                        versus: -1
                                                    }
                                                })
                                            }>▼</div>
                                        </div>
                                    )
                                }</div>
                            </div>
                        </th>
                    ))
                }</tr>
            </thead>

            <tbody className="TableBody tablebody" style={{ maxHeight: height }}>  
                {rows.length ? rows.map((row, i) => (
                    <tr key={`r${i}`}
                        className={`TableRow ${activeRow === row._ID ? (crossHighlight || rowHighlight) : ''}`}
                        style={{ verticalAlign: rowVerticalAlign || 'top' }}
                        onClick={e => {
                            rowClick.call(e, e, row)
                        }}
                        onMouseEnter={e => {
                            rowEnter.call(e, e, row)
                        }}
                        onMouseLeave={e => {
                            rowLeave.call(e, e, row)
                        }}
                    >{
                            columns.map((col, j) => (
                                <td key={`r${i}c${j}`}
                                    align={col.align || 'left'}
                                    className={`tablecell TableCell ${activeCol === col.key ? (crossHighlight || columnHighlight) : ''} ${(cellHightlight && activeRow === row._ID && activeCol === col.key) ? cellHightlight : ''}`}
                                    style={col.width && { width: col.width }}
                                    onClick={e => {
                                        col.onClick && col.onClick.call(e, e, col, row);
                                        cellClick.call(e, e, col, row);
                                        columnClick.call(e, e, col, row);
                                    }}
                                    onMouseEnter={e => {
                                        dispatch({
                                            type: 'cellHover',
                                            payload: {
                                                row,
                                                col
                                            }
                                        });
                                        cellEnter.call(e, e, row);
                                        columnEnter.call(e, e, col);
                                    }}
                                    onMouseLeave={e => {
                                        dispatch({
                                            type: 'cellOut',
                                            payload: {
                                                row,
                                                col
                                            }
                                        });
                                        cellLeave.call(e, e, row);
                                        columnLeave.call(e, e, col);
                                    }}
                                >{
                                        'component' in col
                                            ? col.component(row, col.key)
                                            : row[col.key] || 'none'
                                }</td>
                            ))
                        }</tr>
                )) : <tr
                    className='TableRow'
                    style={{ verticalAlign: 'middle', textAlign:'center' }}
                ><td className='tablecell TableCell' span={columns.length}>{noFilterData}</td></tr>}
                
            </tbody>



            {columns.some(column => 'footerComponent' in column || 'footerLabel' in column) &&
                <tfoot className="TableFooter">
                    <tr>{
                        columns.map((col, k) => (
                            <th key={`h${k}`}
                                className="TableFooterCell"
                                style={col.width && { width: col.width }}
                            >{
                                    'footerComponent' in col
                                        ? col.footerComponent(col, 'key')
                                        : col.footerLabel
                                }</th>
                        ))
                    }</tr>
                </tfoot>
            }
        </table>
        { /* <pre>{JSON.stringify(state, null, 2)}</pre> */}
    </div>

}

export default HyperTable