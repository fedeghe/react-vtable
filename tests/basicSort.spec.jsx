
/**
 * @jest-environment jsdom
 */
import React from "react";
import {
    render,
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event"
import { configure } from "@testing-library/dom";

import HyperTable from "../dist/index.js";
import config from "./configs/basic";

configure({
    testIdAttribute: "data-uie",
});

const Sorter =  ({sort}) => {
    const {sortAsc, sortDesc, unSort, direction, isSorting} = sort;
    return <div>
        <span title={direction === 'asc' ? "sorting": ""} data-uie="sortAsc" onClick={() => sortAsc()}>▲</span>
        {direction && isSorting && <span  data-uie="unsort" onClick={unSort}>&times;</span>}
        <span title={direction === 'desc' ? "sorting" : ""} data-uie="sortDesc" onClick={() => sortDesc()}>▼</span>
    </div>;
};

config.columns[0].header = ({
    column, columnIndex,
    sort, //: {sortAsc, sortDesc, unSort, direction} = {}
    isSorting
}) => <>
    <span>{column.key + '_' + columnIndex}</span>
    <Sorter {...{ column, columnIndex, sort, isSorting }} />
</>;
config.columns[0].sort = ({rowA, rowB, columnKey, direction}) => {
    const v = rowA[columnKey] > rowB[columnKey] ? 1 : -1;
    return {
        asc : v,
        desc: -v
    }[direction];
};
config.header = {height:80};

describe("HyperTable - basic sort", () => {
    it("should behave as expected", () => {
        const { container, getByTestId } = render(
            <HyperTable config={config} />
        );

        expect(container).toMatchSnapshot();
        const sortDesc = getByTestId(`sortDesc`);
        userEvent.click(sortDesc);
        expect(container).toMatchSnapshot();
    });
});
