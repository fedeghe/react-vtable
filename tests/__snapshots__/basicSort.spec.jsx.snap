// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HyperTable - basic sort should behave as expected 1`] = `
<div>
  <div
    class="Wrapper-0-2-1 Wrapper-d0-0-2-2 "
  >
    <div
      class="TableContainer-0-2-7 TableContainer-d0-0-2-14"
    >
      <table
        class="Table-0-2-8"
      >
        <thead
          class="Thead-0-2-16 Thead-d0-0-2-21"
        >
          <tr>
            <th
              class="TableHeader TheadTh-0-2-20 "
            >
              <div
                style="width: 300px;"
              >
                <span>
                  id_0
                </span>
                <div>
                  <span
                    data-uie="sortAsc"
                    title=""
                  >
                    ▲
                  </span>
                  <span
                    data-uie="sortDesc"
                    title=""
                  >
                    ▼
                  </span>
                </div>
              </div>
            </th>
            <th
              class="TableHeader TheadTh-0-2-20 "
            >
              <div
                style="width: 150px;"
              >
                entityid
              </div>
            </th>
            <th
              class="TableHeader TheadTh-0-2-20 "
            >
              <div
                style="width: 300px;"
              >
                name
              </div>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="Tr-0-2-28 Tr-d0-0-2-30"
          >
            <td
              class="Td-0-2-29 Td-d1-0-2-31"
              colspan="3"
            />
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d0-0-2-34"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-0-0"
                >
                  666
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d1-0-2-35"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-0-1"
                >
                  ID0
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d2-0-2-36"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-0-2"
                >
                  a random, string 0
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d3-0-2-37"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-1-0"
                >
                  667
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d4-0-2-38"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-1-1"
                >
                  ID1
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d5-0-2-39"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-1-2"
                >
                  a random, string 1
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d6-0-2-40"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-2-0"
                >
                  668
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d7-0-2-41"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-2-1"
                >
                  ID2
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d8-0-2-42"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-2-2"
                >
                  a random, string 2
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d9-0-2-43"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-3-0"
                >
                  669
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d10-0-2-44"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-3-1"
                >
                  ID3
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d11-0-2-45"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-3-2"
                >
                  a random, string 3
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d12-0-2-46"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-4-0"
                >
                  670
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d13-0-2-47"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-4-1"
                >
                  ID4
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d14-0-2-48"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-4-2"
                >
                  a random, string 4
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d15-0-2-49"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-5-0"
                >
                  671
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d16-0-2-50"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-5-1"
                >
                  ID5
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d17-0-2-51"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-5-2"
                >
                  a random, string 5
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d18-0-2-52"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-6-0"
                >
                  672
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d19-0-2-53"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-6-1"
                >
                  ID6
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d20-0-2-54"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-6-2"
                >
                  a random, string 6
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d21-0-2-55"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-7-0"
                >
                  673
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d22-0-2-56"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-7-1"
                >
                  ID7
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d23-0-2-57"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-7-2"
                >
                  a random, string 7
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d24-0-2-58"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-8-0"
                >
                  674
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d25-0-2-59"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-8-1"
                >
                  ID8
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d26-0-2-60"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-8-2"
                >
                  a random, string 8
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d27-0-2-61"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-9-0"
                >
                  675
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d28-0-2-62"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-9-1"
                >
                  ID9
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d29-0-2-63"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-9-2"
                >
                  a random, string 9
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d30-0-2-64"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-10-0"
                >
                  676
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d31-0-2-65"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-10-1"
                >
                  ID10
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d32-0-2-66"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-10-2"
                >
                  a random, string 10
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d33-0-2-67"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-11-0"
                >
                  677
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d34-0-2-68"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-11-1"
                >
                  ID11
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d35-0-2-69"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-11-2"
                >
                  a random, string 11
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d36-0-2-70"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-12-0"
                >
                  678
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d37-0-2-71"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-12-1"
                >
                  ID12
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d38-0-2-72"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-12-2"
                >
                  a random, string 12
                </div>
              </div>
            </td>
          </tr>
          <tr
            class="Tr-0-2-28 Tr-d2-0-2-73"
          >
            <td
              class="Td-0-2-29 Td-d3-0-2-74"
              colspan="3"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`HyperTable - basic sort should behave as expected 2`] = `
<div>
  <div
    class="Wrapper-0-2-1 Wrapper-d0-0-2-2 "
  >
    <div
      class="TableContainer-0-2-7 TableContainer-d0-0-2-14"
    >
      <table
        class="Table-0-2-8"
      >
        <thead
          class="Thead-0-2-16 Thead-d0-0-2-21"
        >
          <tr>
            <th
              class="TableHeader TheadTh-0-2-20 "
            >
              <div
                style="width: 300px;"
              >
                <span>
                  id_0
                </span>
                <div>
                  <span
                    data-uie="sortAsc"
                    title=""
                  >
                    ▲
                  </span>
                  <span
                    data-uie="unsort"
                  >
                    ×
                  </span>
                  <span
                    data-uie="sortDesc"
                    title="sorting"
                  >
                    ▼
                  </span>
                </div>
              </div>
            </th>
            <th
              class="TableHeader TheadTh-0-2-20 "
            >
              <div
                style="width: 150px;"
              >
                entityid
              </div>
            </th>
            <th
              class="TableHeader TheadTh-0-2-20 "
            >
              <div
                style="width: 300px;"
              >
                name
              </div>
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="Tr-0-2-28 Tr-d0-0-2-30"
          >
            <td
              class="Td-0-2-29 Td-d1-0-2-31"
              colspan="3"
            />
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d39-0-2-83"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-0-0"
                >
                  1665
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d40-0-2-84"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-0-1"
                >
                  ID999
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d41-0-2-85"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-0-2"
                >
                  a random, string 999
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d42-0-2-86"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-1-0"
                >
                  1664
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d43-0-2-87"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-1-1"
                >
                  ID998
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d44-0-2-88"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-1-2"
                >
                  a random, string 998
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d45-0-2-89"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-2-0"
                >
                  1663
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d46-0-2-90"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-2-1"
                >
                  ID997
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d47-0-2-91"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-2-2"
                >
                  a random, string 997
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d48-0-2-92"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-3-0"
                >
                  1662
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d49-0-2-93"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-3-1"
                >
                  ID996
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d50-0-2-94"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-3-2"
                >
                  a random, string 996
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d51-0-2-95"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-4-0"
                >
                  1661
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d52-0-2-96"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-4-1"
                >
                  ID995
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d53-0-2-97"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-4-2"
                >
                  a random, string 995
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d54-0-2-98"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-5-0"
                >
                  1660
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d55-0-2-99"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-5-1"
                >
                  ID994
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d56-0-2-100"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-5-2"
                >
                  a random, string 994
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d57-0-2-101"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-6-0"
                >
                  1659
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d58-0-2-102"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-6-1"
                >
                  ID993
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d59-0-2-103"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-6-2"
                >
                  a random, string 993
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d60-0-2-104"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-7-0"
                >
                  1658
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d61-0-2-105"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-7-1"
                >
                  ID992
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d62-0-2-106"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-7-2"
                >
                  a random, string 992
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d63-0-2-107"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-8-0"
                >
                  1657
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d64-0-2-108"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-8-1"
                >
                  ID991
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d65-0-2-109"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-8-2"
                >
                  a random, string 991
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d66-0-2-110"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-9-0"
                >
                  1656
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d67-0-2-111"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-9-1"
                >
                  ID990
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d68-0-2-112"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-9-2"
                >
                  a random, string 990
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d69-0-2-113"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-10-0"
                >
                  1655
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d70-0-2-114"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-10-1"
                >
                  ID989
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d71-0-2-115"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-10-2"
                >
                  a random, string 989
                </div>
              </div>
            </td>
          </tr>
          <tr
            class=""
          >
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d72-0-2-116"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-11-0"
                >
                  1654
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d73-0-2-117"
                style="width: 150px;"
              >
                <div
                  data-uie="uie-11-1"
                >
                  ID988
                </div>
              </div>
            </td>
            <td
              class="AlTop-0-2-23   "
            >
              <div
                class="Cell-0-2-32 Cell-d74-0-2-118"
                style="width: 300px;"
              >
                <div
                  data-uie="uie-11-2"
                >
                  a random, string 988
                </div>
              </div>
            </td>
          </tr>
          <tr
            class="Tr-0-2-28 Tr-d2-0-2-73"
          >
            <td
              class="Td-0-2-29 Td-d3-0-2-74"
              colspan="3"
            />
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
